cranelift/codegen/src/machinst/abi.rs

```diff
@@ -591,52 +643,76 @@ impl SigData {
 
         // Compute args and retvals from signature. Handle retvals first,
         // because we may need to add a return-area arg to the args.
-        let (rets, sized_stack_ret_space, _) = M::compute_arg_locs(
+        let rets_start = u32::try_from(sigs.abi_args.len()).unwrap();
+        let (sized_stack_ret_space, _) = M::compute_arg_locs(
             sig.call_conv,
             flags,
             returns,
             ArgsOrRets::Rets,
             /* extra ret-area ptr = */ false,
+            ArgsAccumulator::new(&mut sigs.abi_args),
         )?;
+        let rets_end = u32::try_from(sigs.abi_args.len()).unwrap();
+
         let need_stack_return_area = sized_stack_ret_space > 0;
-        let (args, sized_stack_arg_space, stack_ret_arg) = M::compute_arg_locs(
+        let args_start = u32::try_from(sigs.abi_args.len()).unwrap();
+        let (sized_stack_arg_space, stack_ret_arg) = M::compute_arg_locs(
             sig.call_conv,
             flags,
             &sig.params,
             ArgsOrRets::Args,
             need_stack_return_area,
+            ArgsAccumulator::new(&mut sigs.abi_args),
         )?;
+        let args_end = u32::try_from(sigs.abi_args.len()).unwrap();
+
+        let arg_indices = args_start..args_end;
+        let ret_indices = rets_start..rets_end;
 
         trace!(
             "ABISig: sig {:?} => args = {:?} rets = {:?} arg stack = {} ret stack = {} stack_ret_arg = {:?}",
             sig,
-            args,
-            rets,
+            arg_indices,
+            ret_indices,
             sized_stack_arg_space,
             sized_stack_ret_space,
             stack_ret_arg,
         );
 
         Ok(SigData {
-            args,
-            rets,
+            arg_indices,
+            ret_indices,
             sized_stack_arg_space,
             sized_stack_ret_space,
             stack_ret_arg,
             call_conv: sig.call_conv,
         })
     }
 
+    /// Get this signature's ABI arguments.
+    pub fn args<'a>(&self, sigs: &'a SigSet) -> &'a [ABIArg] {
+        let start = usize::try_from(self.arg_indices.start).unwrap();
```

Comment:
Huh, I'm surprised there's no `From<u32> for usize`; I guess this is in case we have 16-bit pointers...

Reply:
Indeed. There is at least one tier 3 target with 16bit pointers.